Add ZoeDepth

NielsRogge opened the pull request:
# What does this PR do?

This PR adds ZoeDepth as introduced in [ZoeDepth: Zero-shot Transfer by Combining Relative and Metric Depth](https://arxiv.org/abs/2302.12288).

To do:

- [x] double check image processor (not sure we can support the [same resize](https://github.com/isl-org/ZoeDepth/blob/edb6daf45458569e24f50250ef1ed08c015f17a7/zoedepth/models/base_models/midas.py#L173)). Update image processor accordingly
- [x] remove testing scripts
- [x] verify relative position bias table/index when loading a beit model from the hub
- [x] add slow integration test
- [x] add image processor tests
- [x] should we add `backbone_hidden_size`?
- [x] make doc tests pass

NielsRogge committed:
First draft

NielsRogge committed:
Fix merge

NielsRogge committed:
Add docs

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_zoedepth

NielsRogge committed:
Clean up code

NielsRogge committed:
Convert model

NielsRogge committed:
Add image processor

NielsRogge committed:
Convert Zoe_K

NielsRogge committed:
More improvements

NielsRogge committed:
Improve variable names and docstrings

NielsRogge committed:
Improve variable names

NielsRogge committed:
Improve variable names

NielsRogge committed:
Replace nn.sequential

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_zoedepth

NielsRogge committed:
More improvements

NielsRogge committed:
Convert ZoeD_NK

NielsRogge committed:
Fix most tests

NielsRogge committed:
Verify pixel values

NielsRogge committed:
Verify pixel values

NielsRogge committed:
Add squeeze

NielsRogge committed:
Update beit to support arbitrary window sizes

NielsRogge committed:
Improve image processor

NielsRogge committed:
Improve docstring

NielsRogge committed:
Improve beit

NielsRogge committed:
Improve model outputs

NielsRogge committed:
Add figure

NielsRogge committed:
Fix beit

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_30136). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

NielsRogge committed:
Update checkpoint

NielsRogge committed:
Fix repo id

NielsRogge committed:
Add _keys_to_ignore_on_load_unexpected

NielsRogge committed:
More improvements

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+    def __init__(self, config, channels, factor):
+        super().__init__()
+        # projection
+        hidden_size = config.backbone_config.hidden_size
```
Same comment here - we can't rely on there being a backbone_config set

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+            Pixel values. Pixel values can be obtained using [`AutoImageProcessor`]. See [`DPTImageProcessor.__call__`]
+            for details.
+
+        head_mask (`torch.FloatTensor` of shape `(num_heads,)` or `(num_layers, num_heads)`, *optional*):
```
head_mask isn't an input

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+        super().__init__()
+        self.K = n_classes
+        self.act = act
+        self.register_buffer("k_idx", torch.arange(0, n_classes).view(1, -1, 1, 1))
+        self.register_buffer("K_minus_1", torch.Tensor([self.K - 1]).view(1, -1, 1, 1))
```
```suggestion
        self.register_buffer("k_idx", torch.arange(0, n_classes).view(1, -1, 1, 1), persistent=False)
        self.register_buffer("K_minus_1", torch.Tensor([self.K - 1]).view(1, -1, 1, 1), persistent=False)
```

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+        super().__init__()
+        self.K = n_classes
+        self.act = act
+        self.register_buffer("k_idx", torch.arange(0, n_classes).view(1, -1, 1, 1))
```
Let's be consistent with capitalization of `k`

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+        y = (
+            log_binom(self.K_minus_1, self.k_idx)
+            + self.k_idx * torch.log(probabilities)
+            + (self.K - 1 - self.k_idx) * torch.log(one_minus_probabilities)
```
why `self.K - 1` and not `self.K_minus_1` here?

amyeroberts commented on src/transformers/models/zoedepth/image_processing_zoedepth.py:
```diff
@@ -0,0 +1,432 @@
+            If `True`, the image is resized to the largest possible size such that the aspect ratio is preserved. Can
+            be overidden by `keep_aspect_ratio` in `preprocess`.
+        ensure_multiple_of (`int`, *optional*, defaults to 32):
+            If `do_resize` is `True`, the image is resized to a size that is a multiple of this value. Can be overidden
+            by `ensure_multiple_of` in `preprocess`.
```
This could be clarified - is `ensure_multiple_of` is set to any numerical value, is `keep_aspect_ratio` always ignored ?

amyeroberts commented on src/transformers/models/zoedepth/image_processing_zoedepth.py:
Tests should be added for the image processor - in particular to make sure the correct image sizes are calculated

NielsRogge replied:
Added tests. cc @ydshieh would it be possible to update `make fixup` to complain if an image processor is added but no tests are present (similar to models, tokenizers)?

amyeroberts commented on src/transformers/models/zoedepth/image_processing_zoedepth.py:
```diff
@@ -0,0 +1,432 @@
+            **kwargs,
+        )
+
+    def pad(
```
Because of different behaviours for `pad` in the feature extractors, the method for padding a single image has been standardized to `pad_image`

```suggestion
    def pad_image(
```

amyeroberts commented on src/transformers/models/zoedepth/image_processing_zoedepth.py:
```diff
@@ -0,0 +1,432 @@
+            image (`np.ndarray`):
+                Image to pad.
+            mode (`PaddingMode`):
+            The padding mode to use. Can be one of:
```
```suggestion
                The padding mode to use. Can be one of:
```

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+            hidden_states = self.reassemble_stage(hidden_states, patch_height, patch_width)
+
+        features = [self.convs[i](feature) for i, feature in enumerate(hidden_states)]
+        # we need the last feature of `features`
```
What does this comment related to - we don't seem to be selection just the last feature

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+        # fusion blocks
+        output = self.fusion_stage(features)
+
+        # we need the last 4 features of `output` as well
```
Same here?

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+        Args:
+            n_classes (`int`, *optional*, defaults to 256):
+                Number of output classes.
+            act (`torch.nn.Module`, *optional*, defaults to `torch.softmax`):
+                Activation function to apply to the output.
```
We shouldn't be applying an activation to a class which calculates a distribution, this transformation should happen separately

NielsRogge replied:
Renamed the class

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+        min_temp=1e-7,
+        act=torch.softmax,
+    ):
+        """Conditional Log Binomial distribution.
```
If it's being passed through an MLP, it's not calculation a distribution here

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+
+        # change to S,N,E format required by transformer
+        embeddings = embeddings.permute(2, 0, 1)
+        S, N, E = embeddings.shape
```
No single letter var names

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+
+    def forward(self, x, prev_bin, prev_bin_embedding=None, interpolate=True):
+        """
+        Args:
```
Should have a description of what this method does

NielsRogge replied:
Done

amyeroberts replied:
This still wasn't resolved

amyeroberts replied:
This still wasn't resolved?

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+        else:
+            delta_c = torch.zeros_like(bin_centers, device=bin_centers.device)
+            for i in range(self.n_attractors):
+                delta_c += dist(
+                    attractors[:, i, ...].unsqueeze(1) - bin_centers
+                )  # shape batch_size, num_bins, height, width
```
```suggestion
                # shape batch_size, num_bins, height, width
                delta_c += dist(attractors[:, i, ...].unsqueeze(1) - bin_centers)
```

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+        bottleneck = out[1]
+        feature_blocks = out[2:]
+
+        x_d0 = self.conv2(bottleneck)
+        x = x_d0
```
lets use better var names here

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+        bottleneck = out[1]
+        feature_blocks = out[2:]
+
+        x_d0 = self.conv2(bottleneck)
+        x = x_d0
```
same here

amyeroberts reviewed (commented):
Thanks for adding this model! 

If we're making changes to BEIT and data2vec vision like this, all of those models slow and doctests should be run to validate this isn't a breaking change

NielsRogge committed:
Address comments

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_zoedepth

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge commented on src/transformers/models/zoedepth/configuration_zoedepth.py:
```diff
@@ -0,0 +1,218 @@
+        attractor_gamma (`int`, *optional*, defaults to 2):
+            The gamma value to use in the attractor.
+        attractor_kind (`str`, *optional*, defaults to `"mean"`):
+            The kind of attractor to use. Can be one of [`"mean"`, `"sum"`].
```
Added at init

NielsRogge committed:
Address comments

NielsRogge commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+"""
+
+
+class ZoeDepthNeck(nn.Module):
```
Moved it

NielsRogge commented on docs/source/en/model_doc/zoedepth.md:
```diff
@@ -0,0 +1,48 @@
+
+## Overview
+
+The ZoeDepth model was proposed in [ZoeDepth: Zero-shot Transfer by Combining Relative and Metric Depth](https://arxiv.org/abs/2302.12288) by Shariq Farooq Bhat, Reiner Birkl, Diana Wofk, Peter Wonka, Matthias Müller. ZoeDepth extends the [DPT](dpt) framework for metric (also called absolute) depth estimation, obtaining state-of-the-art results.
```
That's what the sentence "ZoeDepth extends the [DPT](dpt) framework for metric (also called absolute) depth estimation, obtaining state-of-the-art results." does

NielsRogge committed:
Rename variable name

NielsRogge committed:
Add backbone_hidden_size

NielsRogge committed:
Vectorize

NielsRogge committed:
Vectorize more

NielsRogge commented on src/transformers/models/zoedepth/image_processing_zoedepth.py:
```diff
@@ -0,0 +1,432 @@
+    input_data_format: Optional[Union[str, ChannelDimension]] = None,
+) -> Tuple[int, int]:
+    def constraint_to_multiple_of(val, multiple, min_val=0, max_val=None):
+        x = round(val / multiple) * multiple
```
The original implementation uses `np.round`: https://github.com/isl-org/ZoeDepth/blob/edb6daf45458569e24f50250ef1ed08c015f17a7/zoedepth/models/base_models/midas.py#L100.

NielsRogge committed:
Address comments

NielsRogge committed:
Clarify docstring

NielsRogge committed:
Remove backbone_hidden_size

NielsRogge commented on src/transformers/models/zoedepth/convert_zoedepth_to_hf.py:
```diff
@@ -0,0 +1,427 @@
+
+
+# here we list all keys to be renamed (original name on the left, our name on the right)
+def create_rename_keys(config, model_name):
```
That's how I initially converted models, by writing a function that renames keys (as done for ConvNeXT here for instance: https://github.com/huggingface/transformers/blob/main/src/transformers/models/convnext/convert_convnext_to_pytorch.py#L82). However this usually is more difficult to debug for me, as some names occur in multiple keys, so the last 20 models I've contributed I switched to using `create_rename_keys`, which makes it more easy to manage and see which keys still need to be converted

NielsRogge replied:
I completely rewrote the conversion script, as seen in this commit: https://github.com/NielsRogge/transformers/commit/5febc7fe2d843300b4d3cc914bd4f4a277aef6f3 and subsequent ones

NielsRogge replied:
The `create_rename_keys` function also does not exist anymore.

ydshieh commented:
> Added tests. cc @ydshieh would it be possible to update make fixup to complain if an image processor is added but no tests are present (similar to models, tokenizers)?

(https://github.com/huggingface/transformers/pull/30136#discussion_r1573721245)

@NielsRogge I can take a look on this!

NielsRogge committed:
Fix image processor

NielsRogge committed:
Remove print statements

NielsRogge committed:
Remove print statement

NielsRogge committed:
Fix merge

NielsRogge committed:
Add integration test

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+        prev_bin = nn.functional.interpolate(prev_bin, (height, width), mode="bilinear", align_corners=True)
+        bin_centers = prev_bin
+
+        # note: only attractor_type = "exp" is supported here, since no checkpoints were released with other attractor types
+        distribution = inv_attractor
+
+        if not self.memory_efficient:
+            func = {"mean": torch.mean, "sum": torch.sum}[self.kind]
+            # shape (batch_size, num_bins, height, width)
+            delta_c = func(distribution(attractors_normed.unsqueeze(2) - bin_centers.unsqueeze(1)), dim=1)
+        else:
+            delta_c = torch.zeros_like(bin_centers, device=bin_centers.device)
+            for i in range(self.n_attractors):
+                # shape (batch_size, num_bins, height, width)
+                delta_c += distribution(attractors_normed[:, i, ...].unsqueeze(1) - bin_centers)
```
Same here - why this renaming? 

```suggestion
        if not self.memory_efficient:
            func = {"mean": torch.mean, "sum": torch.sum}[self.kind]
            # shape (batch_size, num_bins, height, width)
            delta_c = func(inv_attractor(attractors_normed.unsqueeze(2) - bin_centers.unsqueeze(1)), dim=1)
        else:
            delta_c = torch.zeros_like(bin_centers, device=bin_centers.device)
            for i in range(self.n_attractors):
                # shape (batch_size, num_bins, height, width)
                delta_c += inv_attractor(attractors_normed[:, i, ...].unsqueeze(1) - bin_centers)
```

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+        prev_bin = nn.functional.interpolate(prev_bin, (height, width), mode="bilinear", align_corners=True)
+        bin_centers = prev_bin
+
+        dist = inv_attractor
+
+        if not self.memory_efficient:
+            func = {"mean": torch.mean, "sum": torch.sum}[self.kind]
+            # shape batch_size, num_bins, height, width
+            delta_c = func(dist(attractors.unsqueeze(2) - bin_centers.unsqueeze(1)), dim=1)
+        else:
+            delta_c = torch.zeros_like(bin_centers, device=bin_centers.device)
+            for i in range(self.n_attractors):
+                # shape batch_size, num_bins, height, width
+                delta_c += dist(attractors[:, i, ...].unsqueeze(1) - bin_centers)
```
same here - just use inv_attractor directly

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+
+        height, width = attractors.shape[-2:]
+
+        prev_bin = nn.functional.interpolate(prev_bin, (height, width), mode="bilinear", align_corners=True)
+        bin_centers = prev_bin
```
same here
```suggestion
        bin_centers = nn.functional.interpolate(prev_bin, (height, width), mode="bilinear", align_corners=True)
```

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+        eps = 1e-3
+        bins = bins + eps
+        bin_widths_normed = bins / bins.sum(dim=1, keepdim=True)
+        bin_widths = (
+            self.max_depth - self.min_depth
+        ) * bin_widths_normed  # shape (batch_size, num_channels, height, width)
```
comments above lines avoids line splitting 

```suggestion
        # shape (batch_size, num_channels, height, width)
        bin_widths = (self.max_depth - self.min_depth) * bin_widths_normed
```

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+        Returns:
+            torch.Tensor: Positional encodings.
+        """
+        position = torch.arange(0, sequence_length, dtype=torch.float32, device=device).unsqueeze(1)
+        index = torch.arange(0, embedding_dim, 2, dtype=torch.float32, device=device).unsqueeze(0)
```
Is having float32 here compatible with loading the model in and using fp16 inputs?

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+        """
+        embeddings = self.embedding_convPxP(x).flatten(2)  # .shape = n,c,s = n, embedding_dim, s
+        if self.use_class_token:
+            # extra special token at start ?
```
🤷‍♀️

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+        Returns:
+            torch.Tensor - Transformer output embeddings of shape (batch_size, sequence_length, embedding_dim)
+        """
+        embeddings = self.embedding_convPxP(x).flatten(2)  # .shape = n,c,s = n, embedding_dim, s
```
clear var names in comment

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+            last_in, bin_embedding_dim, n_classes=n_bins, min_temp=min_temp, max_temp=max_temp
+        )
+
+    def forward(self, out, rel_depth):
+        outconv_activation = out[0]
+        bottleneck = out[1]
+        feature_blocks = out[2:]
```
Rather than pass the inputs as a tuple like this, it's better to hae `outconv_activation`, `bottleneck` and `feature_blocks` are input arguments to the forward method. Same goes for `ZoeDepthMultipleMetricDepthEstimationHeads`

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+        x = self.conv2(bottleneck)
+        _, seed_bin_centers = self.seed_bin_regressor(x)
+
+        if self.bin_centers_type == "normed" or self.bin_centers_type == "hybrid2":
```
```suggestion
        if self.bin_centers_type in ("normed", "hybrid2"):
```

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+        )
+
+        if self.use_batch_norm:
+            self.batch_norm1 = nn.BatchNorm2d(config.fusion_hidden_size)
+            self.batch_norm2 = nn.BatchNorm2d(config.fusion_hidden_size)
```
batch norm layers should use `batch_norm_eps` to instanstiate themselves

amyeroberts commented on src/transformers/models/zoedepth/image_processing_zoedepth.py:
```diff
@@ -0,0 +1,448 @@
+        torch_image = torch_image.permute(0, 3, 1, 2) if input_data_format == "channels_last" else torch_image
+
+        # TODO support align_corners=True in image_transforms.resize
+        resample_to_mode = {PILImageResampling.BILINEAR: "bilinear", PILImageResampling.BICUBIC: "bicubic"}
+        mode = resample_to_mode[resample]
+        resized_image = nn.functional.interpolate(
+            torch_image, (int(height), int(width)), mode=mode, align_corners=True
+        )
+        resized_image = (
+            resized_image.squeeze().permute(1, 2, 0).numpy()
+            if data_format == "channels_last"
+            else resized_image.squeeze().numpy()
+        )
```
if torch is used here - it's necessary to add protections around it e.g. `requires_backends`

amyeroberts commented on src/transformers/models/zoedepth/image_processing_zoedepth.py:
```diff
@@ -0,0 +1,448 @@
+        torch_image = torch.from_numpy(image).unsqueeze(0)
+        torch_image = torch_image.permute(0, 3, 1, 2) if input_data_format == "channels_last" else torch_image
+
+        # TODO support align_corners=True in image_transforms.resize
```
An issue should be opened. Did you look at converting the c implementation?

NielsRogge replied:
Opened an issue here: https://github.com/huggingface/transformers/issues/30525.

amyeroberts commented on src/transformers/models/beit/modeling_beit.py:
```diff
@@ -269,7 +272,8 @@ def forward(
         # Add relative position bias if present.
         if self.relative_position_bias is not None:
-            attention_scores = attention_scores + self.relative_position_bias().unsqueeze(0)
+            window_size = tuple(np.array(resolution) // self.config.patch_size)
```
Why are we converting to a numpy array and then tuple here?

amyeroberts commented on src/transformers/models/beit/modeling_beit.py:
```diff
@@ -455,6 +486,16 @@ def __init__(self, config: BeitConfig, window_size: tuple) -> None:
+        self.register_buffer("relative_position_index", relative_position_index, persistent=False)
+        self.relative_position_indices = {}
+
+    def generate_relative_position_index_original(self, window_size: Tuple[int, int]):
```
Having these two methods is a good indication that we shouldn't be doing what we're doing. The modeling file shouldn't become a place to keep multiple possible code paths. If adding zoedepth means the beit model is affected substantially, then we should think about other ways to integrate the model

amyeroberts commented on src/transformers/models/beit/modeling_beit.py:
```diff
@@ -472,15 +513,67 @@ def __init__(self, config: BeitConfig, window_size: tuple) -> None:
+
+        relative_position_bias = new_relative_position_bias_table[self.relative_position_indices[key].view(-1)].view(
+            window_size[0] * window_size[1] + 1, window_size[0] * window_size[1] + 1, -1
+        )  # Wh*Ww,Wh*Ww,nH
```
* clearer var names in comment
* put comment above the code of interest
* split the code across lines instead of trying to do a one-liner which then line splits

amyeroberts commented on src/transformers/models/beit/modeling_beit.py:
```diff
@@ -472,15 +513,67 @@ def __init__(self, config: BeitConfig, window_size: tuple) -> None:
+        relative_position_bias = new_relative_position_bias_table[self.relative_position_indices[key].view(-1)].view(
+            window_size[0] * window_size[1] + 1, window_size[0] * window_size[1] + 1, -1
+        )  # Wh*Ww,Wh*Ww,nH
+        relative_position_bias = relative_position_bias.permute(2, 0, 1).contiguous()  # nH, Wh*Ww, Wh*Ww
```
same here

amyeroberts reviewed (commented):
Thanks for the continued work on this! 

Overall looking pretty clean, however there's still quite a few formatting areas to address. Main comments are about the changes to beit and classes using the config to instantiate themselves

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge committed:
Address comments

NielsRogge committed:
Add requires_backends

NielsRogge committed:
Clean up

NielsRogge committed:
Simplify conversion script

NielsRogge committed:
Simplify more

NielsRogge committed:
Simplify more

NielsRogge committed:
Simplify more

NielsRogge committed:
Clean up

NielsRogge committed:
Make sure beit is loaded correctly

NielsRogge commented:
Thanks for your review. I've addressed all comments, the issue with BEiT is fixed as I now know that one can provide a [regex pattern for unexpected keys to be ignored](https://github.com/huggingface/transformers/blob/73014b561d5f88d728e46a57d346f516fefe3f2d/src/transformers/modeling_utils.py#L1264-L1267) (didn't know this :D ). 

cc @ydshieh there seems to be an issue regarding the new slow CI tests for a new model. Could you help me in looking into this?

NielsRogge commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1347 @@
+        ...     predicted_depth = outputs.predicted_depth
+
+        >>> # interpolate to original size
+        >>> prediction = torch.nn.functional.interpolate(
```
Are you fine with adding a separate `post_process_depth_estimation` model to image processors of depth models (DPT, Depth Anything and ZoeDepth) in a follow-up PR?

NielsRogge commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+            Attractor = ZoeDepthAttractorLayerUnnormed
+
+        self.seed_bin_regressor = SeedBinRegressorLayer(
+            bottleneck_features, n_bins=n_bins, min_depth=min_depth, max_depth=max_depth
```
This is different depending on whether a single metric head or multiple are used. In the multiple heads case, one provides different `n_bins`, `min_depth` and `max_depth` for the different bin configurations.

amyeroberts replied:
I'm open to different ways of tackling this. As it stands though, the current implementation shows a misconfiguration - it shouldn't be necessary to pass in all of these arguments as kwargs and the heads shouldn't be configured with different arguments. For example - why not use bin_configurations in both cases?

NielsRogge commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+        self.seed_bin_regressor = SeedBinRegressorLayer(
+            bottleneck_features, n_bins=n_bins, min_depth=min_depth, max_depth=max_depth
+        )
+        self.seed_projector = ZoeDepthProjector(bottleneck_features, bin_embedding_dim)
```
Same reply here as above, this is different depending on whether a single metric head or multiple are used.

NielsRogge commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1349 @@
+        self.seed_projector = ZoeDepthProjector(bottleneck_features, bin_embedding_dim)
+
+        self.projectors = nn.ModuleList(
+            [ZoeDepthProjector(config.fusion_hidden_size, bin_embedding_dim) for _ in range(4)]
```
Same reply here as above, this is different depending on whether a single metric head or multiple are used.

NielsRogge replied:
In the case of multiple heads, then the `mlp_dim` of `ZoeDepthProjector` is set to `config.bin_embedding_dim // 2` (which is 128 // 2 = 64), whereas in case only a single head is used, then the `mlp_dim` defaults to 128.

ydshieh commented:
I have to check why such error in  Run all tests for the new model (models/zoedepth, single-gpu) (pull_request)

NielsRogge committed:
Address comment

NielsRogge committed:
Address bin_configurations

NielsRogge committed:
Use bin_configurations

NielsRogge committed:
Convert models, add integration tests

NielsRogge committed:
Fix doc test

NielsRogge commented:
@amyeroberts feel free to approve the PR, all checkpoints are on the hub now: https://huggingface.co/models?other=zoedepth

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1331 @@
+            [
+                Attractor(
+                    config,
+                    bin_embedding_dim,
```
Same here re bin embedding dim

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1331 @@
+            ]
+        )
+
+        last_in = N_MIDAS_OUT + 1  # +1 for relative depth
```
N_MIDAS_OUT should be in the config if it's configuring the layers

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1331 @@
+        metric_depth = metric_depth.squeeze(dim=1)
+
+        loss = None
+        if labels is not None:
+            raise NotImplementedError("Training is not implemented yet")
```
Let's put this at the top - we should fail as early as possible i.e. not before we do an expensive forward pass

amyeroberts reviewed (commented):
Thanks for the continued work on this! Still a few loose strands to tidy up and make sure the implementation matches the library patterns. 

Before merging we should make sure that the affected models are still all backwards compatible. To run the slow tests could you: 
* Rebase on main
* Push an empty commit with the command `[run-slow]beit,data2vec,zoedepth`

> @amyeroberts feel free to approve the PR, all checkpoints are on the hub now: https://huggingface.co/models?other=zoedepth

I know you probably mean well, but just to make this process clear: having PRs merged in is a methodical one. A contributor writes code, they request a review, and then the reviewer goes through the code. PRs are not approved on the basis of whether or not the review "feels free" to do so. What determines whether a PR is mergeable is whether the code is good enough; not on the basis of some whim or lack of feeling that they are allowed or able to do so.

NielsRogge committed:
Address comments

NielsRogge committed:
Unify regressor classes

NielsRogge committed:
Clarify arguments

NielsRogge committed:
Improve resize_image

ydshieh commented:
Hi @NielsRogge  You might want to pull the latest main and rebase. It looks like there is an issue of the PR slow CI (couldn't find a file)

NielsRogge committed:
Add num_relative_features

NielsRogge commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1331 @@
+        return self.log_binomial_transform(probabilities, temperature)
+
+
+class ZoeDepthSeedBinRegressor(nn.Module):
```
Added the logic of both in a single `ZoeDepthSeedBinRegressor` class.

However, the Attractor classes have quite some differences between their normed and unnormed versions, so unifying them into a single class would result in many if-else statements, hurting readability. So now we have 2 different versions which is not really consistent

amyeroberts replied:
I would have them in one class in the attractor. There's already a lot of if/else logic in the forward pass, I don't think this would make much of a difference. Up to you depending on whether you care more about the consistency or not

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_zoedepth

NielsRogge committed:
Address comment

NielsRogge committed:
[run-slow]beit,data2vec,zoedepth

NielsRogge commented:
cc @ydshieh looks like there's an issue with the CI runner

ydshieh commented:
Hi @NielsRogge .

Thanks for flagging. At least, all the pytest are passing, it's just artifact upload issue. Will check.

ydshieh commented:
Ah, I see the problem - the same model runs the tests twice!

ydshieh commented:
Fixed #30634

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_zoedepth

NielsRogge committed:
[run-slow]beit,data2vec,zoedepth

NielsRogge commented:
Hi @amyeroberts addressed all comments, CI is green

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1302 @@
+        batch_size, _, height, width = attractors.shape
+        attractors = attractors.view(batch_size, self.n_attractors, 2, height, width)
+        # batch_size, num_attractors, 2, height, width
+        attractors_normed = attractors / attractors.sum(dim=2, keepdim=True)
+        attractors_normed = attractors[:, :, 0, ...]  # batch_size, batch_size*num_attractors, height, width
```
I think there's an error here `attractors_normed` is immediately overwritten by just the unnormed attractors

NielsRogge replied:
This was taken from the [original implementation](https://github.com/isl-org/ZoeDepth/blob/edb6daf45458569e24f50250ef1ed08c015f17a7/zoedepth/models/layers/attractor.py#L105-L106)

NielsRogge replied:
I have updated the code to clarify that this is a bug in the original implementation, and that we keep the same logic in order to make sure logits are the same

amyeroberts commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1302 @@
+        condition_dim,
+        n_classes=256,
+        bottleneck_factor=2,
+        p_eps=1e-4,
+        act=torch.softmax,
```
Arguments which aren't modified / used should be removed as kwargs

amyeroberts commented on src/transformers/models/beit/modeling_beit.py:
```diff
@@ -455,32 +462,71 @@ def __init__(self, config: BeitConfig, window_size: tuple) -> None:
+            [new_sub_table, old_relative_position_bias_table[old_num_relative_distance - 3 :]]
+        )
+
+        key = str(window_size[1]) + "," + str(window_size[0])
```
Why convert to a string here? If window_size[0], window_size[1] are ints, we can just use these to index in the dictionary

amyeroberts reviewed (commented):
Thanks for the continued work on this! 

There's still a few structural pieces to tidy up, but looking close to merge!

NielsRogge committed:
Address comments

NielsRogge committed:
Address comment

NielsRogge committed:
Fix merge

NielsRogge committed:
Address comment

NielsRogge commented on src/transformers/models/zoedepth/modeling_zoedepth.py:
```diff
@@ -0,0 +1,1302 @@
+        """
+        super().__init__()
+        encoder_layers = nn.TransformerEncoderLayer(embedding_dim, num_heads, dim_feedforward=1024, batch_first=True)
+        self.transformer_encoder = nn.TransformerEncoder(encoder_layers, num_layers=4)
```
Ok, although models like SigLIP and Idefics2 use `nn.MultiheadAttention`: https://github.com/huggingface/transformers/blob/c11ac7857bf93914f69a209d50a918ad7fe1d9a3/src/transformers/models/idefics2/modeling_idefics2.py#L511

amyeroberts replied:
Idefics2 uses it because it's using copied siglip modules. Siglip shouldn't have had it in the first place, probably just missed in review.

NielsRogge committed:
Replace nn.TransformerEncoderLayer and nn.TransformerEncoder

NielsRogge committed:
Replace nn.MultiheadAttention

NielsRogge committed:
Add attributes for patch transformer to config

NielsRogge commented on tests/models/zoedepth/test_image_processing_zoedepth.py:
```diff
@@ -0,0 +1,131 @@
+
+    def test_keep_aspect_ratio(self):
+        size = {"height": 512, "width": 512}
+        image_processor = ZoeDepthImageProcessor(size=size, keep_aspect_ratio=True, ensure_multiple_of=32)
```
Yes so the core logic is implemented in [this function](https://github.com/NielsRogge/transformers/blob/7d0e82b62f07579048f21b9a432d8a5f9abd2a05/src/transformers/models/zoedepth/image_processing_zoedepth.py#L52). The `ensure_multiple_of` is always executed at the end.

amyeroberts replied:
This wasn't resolved

amyeroberts replied:
`enusre_multiple_of` should also be tested with `keep_aspect_ratio=False`

NielsRogge committed:
Add tests for ensure_multiple_of

NielsRogge committed:
Update organization

NielsRogge commented on docs/source/en/model_doc/zoedepth.md:
```diff
@@ -0,0 +1,108 @@
+    outputs = model(**inputs)
+    predicted_depth = outputs.predicted_depth
+
+# interpolate to original size
```
Issue has been opened: #30917

amyeroberts reviewed (commented):
Thanks for the work adding this model! Looks nearly ready to merge. 

A few final things to do before: 
* Run slow tests for beit and data2vec. This can be done through pushing an empty commit `[run-slow] beit data2vec`
* Run the doc examples and confirm (with screenshot?) that they're passing 
* Clarify the norming logic 

The final bit that needs to be confirmed/addressed is making sure that this model doesn't suffer from the same issues that have been seen in the DPT model #28292 (as some of this is copied from) i.e. the model shouldn't create weights that never used.

NielsRogge committed:
Add tests

NielsRogge committed:
[run-slow] beit data2vec

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_zoedepth

NielsRogge committed:
Update ruff

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_zoedepth

NielsRogge committed:
[run-slow] beit data2vec

NielsRogge committed:
Add comment

NielsRogge commented:
Ok,

Pushed an empty commit to run the slow tests which confirmed all slow test are passing, the doc tests are tested by the `build_pr_documentation` check right?

amyeroberts commented:
> the doc tests are tested by the build_pr_documentation check right?

No, this checks the documentation can be built i.e. the webpage. 

See: https://huggingface.co/docs/transformers/main/en/testing#run-documentation-tests

NielsRogge commented:
Sorry I meant the `tests_pr_documentation_tests` check

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_zoedepth

amyeroberts commented:
Ah, yes, that should be fine as long as the docs and model examples are tested. The only thing is from the recent CI workflow, it doesn't look like the tests have [properly run (?)](https://app.circleci.com/pipelines/github/huggingface/transformers/94086/workflows/243f65ca-62c3-46f3-8dba-8af883b7d0dd/jobs/1235117/steps). At least, it's not possible to see which tests have run at all there

amyeroberts commented:
@NielsRogge Have you tested for the issues related to weight loading / DPT?

NielsRogge commented:
Yes, can confirm:
```python
>>> from transformers import ZoeDepthForDepthEstimation
>>> model = ZoeDepthForDepthEstimation.from_pretrained("Intel/zoedepth-nyu-kitti")
config.json: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2.23k/2.23k [00:00<00:00, 1.41MB/s]
model.safetensors: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1.38G/1.38G [00:29<00:00, 47.2MB/s]
>>> model = ZoeDepthForDepthEstimation.from_pretrained("Intel/zoedepth-nyu")
config.json: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2.13k/2.13k [00:00<00:00, 3.27MB/s]
model.safetensors: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1.38G/1.38G [00:29<00:00, 46.4MB/s]
>>> model = ZoeDepthForDepthEstimation.from_pretrained("Intel/zoedepth-kitti")
config.json: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2.12k/2.12k [00:00<00:00, 1.52MB/s]
model.safetensors: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1.38G/1.38G [00:29<00:00, 46.0MB/s]
>>> 
```

amyeroberts commented on src/transformers/models/zoedepth/image_processing_zoedepth.py:
```diff
@@ -0,0 +1,444 @@
+            If `True`, the image is resized to the largest possible size such that the aspect ratio is preserved. Can
+            be overidden by `keep_aspect_ratio` in `preprocess`.
+        ensure_multiple_of (`int`, *optional*, defaults to 32):
+            If `do_resize` is `True`, the image is resized to a size that is a multiple of this value. Works both with and without
+            `keep_aspect_ratio` being set to `True`. Can be overidden by `ensure_multiple_of` in `preprocess`.
```
This should be clarified a bit. If `keep_aspect_ratio` is `True`, which one takes precedence i.e will the output image be a mutliple of, or have the same aspect ratio?

amyeroberts reviewed (commented):
Thanks for running all the tests to confirm the model behaviours! 

There's still some comments which weren't resolved. In particular, the behaviour of the image processor's arguments wrt keep_aspect_ratio and ensure_multiple_of need to be throughly tested

NielsRogge committed:
Improve docstrings, add test

NielsRogge committed:
Fix merge

NielsRogge committed:
Fix interpolate_pos_encoding

NielsRogge committed:
Fix slow tests

NielsRogge committed:
Add docstring

NielsRogge commented:
@amyeroberts addressed your comment, failing CI is unrelated.

amyeroberts reviewed (commented):
Thanks for the continued work on this! 

There's still a bit of work do to on the testing and documentation of `keep_aspect_ratio` and `ensure_multiple_of`, otherwise the PR looks good! 

I realise it might seem like I'm being picky here, but there's two reasons why this is important: 
* Users should be able to read the docstring and know how to use the objects. At the moment, the descriptions mean that someone still has to go look at the code. 
* There have been quite a few fixes to resizing logic for models like e.g. yolos. This wasn't caught because the behaviour wasn't properly tested. It's important we test thoroughly and correctly now, as it's harder to fix post-merge. 

In particular, when testing, it's fine if some of the logic seems repetitive. [Tests should be DAMP rather than DRY](https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html). They serve not just as a safety net, but also documentation. As such, it's important we isolate so we're testing a single idea at a time, and the behaviour being tested is obvious (e.g. having the same output with `keep_aspect_ratio` as `True` and `False` doesn't tell us about what `keep_aspect_ratio` does).

NielsRogge committed:
Update src/transformers/models/zoedepth/image_processing_zoedepth.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

NielsRogge committed:
Update src/transformers/models/zoedepth/image_processing_zoedepth.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

NielsRogge committed:
Improve tests and docstrings

amyeroberts commented on tests/models/zoedepth/test_modeling_zoedepth.py:
```diff
@@ -0,0 +1,256 @@
+        self.config_tester = ConfigTester(self, config_class=ZoeDepthConfig, has_text_modality=False, hidden_size=37)
+
+    def test_config(self):
+        self.config_tester.create_and_test_config_to_json_string()
+        self.config_tester.create_and_test_config_to_json_file()
+        self.config_tester.create_and_test_config_from_and_save_pretrained()
+        self.config_tester.create_and_test_config_with_num_labels()
+        self.config_tester.check_config_can_be_init_without_params()
+        self.config_tester.check_config_arguments_init()
```
This should use `self.config_tester.run_common_tests()`

NielsRogge committed:
Fix merge

NielsRogge committed:
Use run_common_tests

NielsRogge committed:
Improve docstrings

NielsRogge committed:
Improve docstrings

NielsRogge committed:
Improve tests

NielsRogge committed:
Improve tests

NielsRogge commented on tests/models/zoedepth/test_image_processing_zoedepth.py:
```diff
@@ -0,0 +1,181 @@
+        self.assertTrue(pixel_values.shape[2] % multiple == 0)
+        self.assertTrue(pixel_values.shape[3] % multiple == 0)
+
+    def test_keep_aspect_ratio(self):
```
Updated the tests to reflect this (turning off all other variables which affect the size in order to test the given variable).

Basically the order of priority is 1) the size argument, which determins the output size 2) the keep_aspect_ratio attribute, which will scale the width and height to the same number, using the smaller of the 2 3) the ensure_multiple_of variable which is used as a final step to round the scale by rounding to the nearest integer.

NielsRogge committed:
Remove print statements

amyeroberts reviewed (approved):
Thanks for all the work adding this model!

NielsRogge committed:
Merge remote-tracking branch 'upstream/main' into add_zoedepth

NielsRogge commented:
Slow tests passing locally, CI failing tests are unrelated, merging.

amyeroberts commented:
@NielsRogge Slow tests have to pass on the CI runs, not locally before merging. Please do not forcibly merge like this again. There are numerical differences which are introduced from the running environment and hardware which mean integration values may not match. You'll need to follow this up and make sure the slow tests pass on our CI images.
